### ISSUE TITLE
Remove dependency on game.dtd

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#1572]
(I'm creating this issue with a description scraped from [this comment](https://github.com/triplea-game/triplea/pull/1566#issuecomment-284910835).  I need a place to start a discussion and document some information I've discovered, and I'd rather do it in a new top-level issue than in response to a comment buried in another issue.  @DanVanAtta, when you have time, please update this description as you see fit and remove these parenthetical remarks.)

A dtd can have value, but in our case I do not see it. It breaks backwards compatibility in the XML files, we can't introduce new tags and have them be ignored, we get errors instead. It also forces an unnecessary ordering on XML tags which should not be enforced. Last, the default values we get from it is scary/bad. It's a non-obvious layer of indirection, the code should be instantiating those values directly. Last, we have to deal with it from a pathing and packaging issue.

Hence, overall I would prefer we would just remove it and do the work necessary to get rid of it. We should have a simpler more straight forward system later, and will then have more freedom to update map xmls, and less issues with map makers manipulating XMLs.

**Comment by ssoloff** [comment]
Below is a table of all attributes that have a default value specified in `game.dtd`:

Element path | Attribute name | Default attribute value
:-- | :-- | :--
game/map/grid | horizontal-connections | "explicit"
&nbsp;  | vertical-connections | "explicit"
&nbsp; | diagonal-connections | "explicit"
game/map/territory | water | "false"
game/playerList/player | optional | "false"
&nbsp; | canBeDisabled | "false"
game/technology/technologies/techname | tech | ""
game/attachmentList/attachment | type | "unitType"
game/attachmentList/attachment/option | count | ""
game/propertyList/property | value | ""
&nbsp; | editable | "false"
&nbsp; | player | ""

**Comment by ssoloff** [comment]
I had originally considered using a tool like [xmllint](http://xmlsoft.org/xmllint.html) to transform all game XML files in the various `triplea-maps` repos to a) explicitly specify a value for all attributes that have a default value specified in the DTD, and b) remove the DTD declaration.  This would preserve the existing `GameParser` logic and would remove all traces of `game.dtd` from the existing maps.

Per game XML file, this can be done with the following command:

```
$ xmllint --valid --dtdattr --dropdtd <file.xml> > tmp.xml; mv tmp.xml <file.xml>
```

where `<file.xml>` is the game XML file to be transformed.  (This command assumes a copy of `game.dtd` has been, at least temporarily, co-located with the game XML file.)  I believe this could be easily automated for all game XML files across all repos.

However, there may be a less intrusive solution that would only require small changes to `GameParser`.  Consider that `GameParser` could be modified to hook requests to load `game.dtd` and instead return an empty stream.  Without the DTD, the underlying XML parser will return an empty string (`""`) in response to an `org.w3c.dom.Element#getAttribute()` call for a missing attribute.  The following table summarizes how `GameParser` would behave in that case:

Attribute path | `GameParser` behavior | Correct?
:-- | :-- | :--
game/map/grid/@horizontal-connections | Parser will throw exception if value is empty string (unsupported value) | ✗
game/map/grid/@vertical-connections | Parser will throw exception if value is empty string (unsupported value) | ✗
game/map/grid/@diagonal-connections | Parser will throw exception if value is empty string (unsupported value) | ✗
game/map/territory/@​water | Parser interprets an empty string the same as if value were "false" | ✓
game/playerList/player/@​optional | Parser interprets an empty string the same as if value were "false" | ✓
game/playerList/player/@​canBeDisabled | Parser interprets an empty string the same as if value were "false" | ✓
game/technology/technologies/techname/@​tech | Parser already handles an empty string | ✓
game/attachmentList/attachment/@​type | Parser will throw exception if value is empty string (unsupported value) | ✗
game/attachmentList/attachment/option/@​count | Parser already handles an empty string | ✓
game/propertyList/property/@​value | Parser already handles an empty string | ✓
game/propertyList/property/@​editable | Parser interprets an empty string the same as if value were "false" | ✓
game/propertyList/property/@​player | Parser does not appear to process this attribute (deprecated?) | ✓

The last column is my interpretation of whether or not `GameParser` would correctly process the attribute value (as an empty string) if the DTD were not loaded to provide the current default values.  As you can see, the majority of the attributes are handled correctly.  The following changes to `GameParser` could be made to allow the remaining attributes to be handled correctly at parity when the DTD was loaded:

Attribute path | `GameParser` changes
:-- | :--
game/map/grid/@horizontal-connections | Treat empty string as if value is "explicit"
game/map/grid/@vertical-connections | Treat empty string as if value is "explicit"
game/map/grid/@diagonal-connections | Treat empty string as if value is "explicit"
game/attachmentList/attachment/@​type | Treat empty string as if value is "unitType"

That is, we're simply moving the default values from the DTD into code, which may be easier to evolve.

In summary, this second solution would only require a code change; existing maps should work as-is.  (However, I would propose that some content in the `triplea-maps/Project` repo be updated to reflect that `game.dtd` is no longer used [e.g. update developer documentation, remove DTD declaration from sample XML files, etc.].)

@DanVanAtta, @ron-murhammer: Any thoughts on which path is preferable?  Or any other alternative I haven't considered?

**Comment by DanVanAtta** [comment]
> "That is, we're simply moving the default values from the DTD into code, which may be easier to evolve."

This  ^   :)


I do find it quite indirect to have values coming from a DTD file and would be happy to see that fixed, and/or the DTD removed entirely.

**Comment by DanVanAtta** [comment]
@ssoloff any objection if we mark this as an ice-box'ed project and close it? It seems like the direction is clear, move the default values from the `dtd` to game engine code. After that we can kill off the `dtd` if we so choose.

**Comment by ssoloff** [comment]
@DanVanAtta Sorry, somehow I missed your comment from June 10.  Yes, with your confirmation, I think the direction is now clear, and I'll try to work on it interleaved with the serialization proxy stuff.
